File: Client/Frontend/Reader/ReadabilityBrowserHelper.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

(function() {
    // To keep this file readable and Readability.js separate, since we import it from an external
    // repository, we merge it in this file programatically. We do not include it as a user script
    // because that means pages can mess with it; by including it below, it is part of an anonymous
    // function that only exists once.

    %READABILITYJS%

    var uri = {
        spec: document.location.href,
        host: document.location.host,
        prePath: document.location.protocol + "//" + document.location.host, // TODO This is incomplete, needs username/password and port
        scheme: document.location.protocol.substr(0, document.location.protocol.indexOf(":")),
        pathBase: document.location.protocol + "//" + document.location.host + location.pathname.substr(0, location.pathname.lastIndexOf("/") + 1)
    }

    // document.cloneNode() can cause the webview to break (bug 1128774).
    // Serialize and then parse the document instead.
    var docStr = new XMLSerializer().serializeToString(document);
    var doc = new DOMParser().parseFromString(docStr, "text/html");

    var readability = new Readability(uri, doc);
    var readabilityResult = readability.parse();

    webkit.messageHandlers.readabilityMessageHandler.postMessage(readabilityResult);
})();

File: Client/Frontend/Reader/ReadabilityBrowserHelper.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import Foundation
import WebKit

protocol ReadabilityBrowserHelperDelegate {
    func readabilityBrowserHelper(readabilityBrowserHelper: ReadabilityBrowserHelper, didFinishWithReadabilityResult result: ReadabilityResult)
}

class ReadabilityBrowserHelper: BrowserHelper {
    var delegate: ReadabilityBrowserHelperDelegate?

    class func name() -> String {
        return "ReadabilityBrowserHelper"
    }

    init?(browser: Browser) {
        if let readabilityPath = NSBundle.mainBundle().pathForResource("Readability", ofType: "js"),
           let readabilitySource = try? NSMutableString(contentsOfFile: readabilityPath, encoding: NSUTF8StringEncoding),
           let readabilityBrowserHelperPath = NSBundle.mainBundle().pathForResource("ReadabilityBrowserHelper", ofType: "js"),
           let readabilityBrowserHelperSource = try? NSMutableString(contentsOfFile: readabilityBrowserHelperPath, encoding: NSUTF8StringEncoding) {
            readabilityBrowserHelperSource.replaceOccurrencesOfString("%READABILITYJS%", withString: readabilitySource as String, options: NSStringCompareOptions.LiteralSearch, range: NSMakeRange(0, readabilityBrowserHelperSource.length))
            let userScript = WKUserScript(source: readabilityBrowserHelperSource as String, injectionTime: WKUserScriptInjectionTime.AtDocumentEnd, forMainFrameOnly: true)
            browser.webView!.configuration.userContentController.addUserScript(userScript)
        }
    }

    func scriptMessageHandlerName() -> String? {
        return "readabilityMessageHandler"
    }

    func userContentController(userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage) {
        if let readabilityResult = ReadabilityResult(object: message.body) {
            delegate?.readabilityBrowserHelper(self, didFinishWithReadabilityResult: readabilityResult)
        }
   }
}